File: test_web_interface.py
#!/usr/bin/env python
"""Test the web interface using Selenium"""

import time
import sys
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException

def test_web_interface():
    print("Starting web interface test...")
    
    # Setup Chrome options
    options = webdriver.ChromeOptions()
    options.add_argument('--headless')  # Run in headless mode
    options.add_argument('--no-sandbox')
    options.add_argument('--disable-dev-shm-usage')
    
    try:
        # Create driver
        driver = webdriver.Chrome(options=options)
        driver.set_window_size(1280, 800)
        
        # Navigate to the app
        print("Navigating to http://localhost:5050...")
        driver.get("http://localhost:5050")
        
        # Wait for page to load
        wait = WebDriverWait(driver, 10)
        new_game_btn = wait.until(EC.presence_of_element_located((By.ID, "new-game-btn")))
        print("✓ Page loaded successfully")
        
        # Test 1: Start a new game
        print("Testing new game...")
        new_game_btn.click()
        time.sleep(0.5)
        
        # Check if board is rendered
        tiles = driver.find_elements(By.CLASS_NAME, "tile")
        print(f"✓ Board rendered with {len(tiles)} tiles")
        
        # Test 2: Check if AI play button is enabled
        ai_play_btn = driver.find_element(By.ID, "ai-play-btn")
        
        if ai_play_btn.is_enabled():
            print("✓ AI Play button is enabled (OCaml backend available)")
            
            # Test 3: Let AI make a few moves
            print("Testing AI moves...")
            ai_play_btn.click()
            
            # Wait for AI to make moves
            inference_time_found = False
            for i in range(10):
                time.sleep(0.2)
                
                # Check inference time
                inference_time_el = driver.find_element(By.ID, "inference-time")
                inference_time = inference_time_el.text
                
                if inference_time != "-":
                    inference_time_found = True
                    print(f"✓ AI move {i+1}: Inference time = {inference_time} ms")
                
                # Check score and max tile
                score = driver.find_element(By.ID, "game-score").text
                max_tile = driver.find_element(By.ID, "game-max-tile").text
                print(f"  Score: {score}, Max tile: {max_tile}")
                
                # Stop after a few moves to keep test quick
                if i >= 5:
                    break
            
            if inference_time_found:
                print("✓ OCaml inference times displayed correctly")
            else:
                print("✗ No inference times found - OCaml backend may not be working")
            
            # Stop AI
            try:
                ai_stop_btn = driver.find_element(By.ID, "ai-stop-btn")
                if ai_stop_btn.is_displayed():
                    ai_stop_btn.click()
            except:
                pass
                
        else:
            print("✗ AI Play button is disabled - OCaml backend not available")
            return False
        
        print("\n✅ All tests passed!")
        return True
        
    except TimeoutException:
        print("❌ Test failed: Page failed to load")
        return False
    except Exception as e:
        print(f"❌ Test failed: {str(e)}")
        return False
    finally:
        driver.quit()

if __name__ == "__main__":
    # Check if server is running
    import requests
    try:
        response = requests.get("http://localhost:5050", timeout=2)
    except:
        print("Error: Flask server is not running on http://localhost:5050")
        print("Please run: cd python && python app.py")
        sys.exit(1)
    
    # Run test
    success = test_web_interface()
    sys.exit(0 if success else 1)

File: test_web_interface.js
const puppeteer = require('puppeteer');

async function testWebInterface() {
    console.log('Starting web interface test...');
    
    // Launch browser
    const browser = await puppeteer.launch({
        headless: false, // Set to true for CI/CD
        args: ['--no-sandbox', '--disable-setuid-sandbox']
    });
    
    try {
        const page = await browser.newPage();
        
        // Set viewport
        await page.setViewport({ width: 1280, height: 800 });
        
        // Go to the app
        console.log('Navigating to http://localhost:5050...');
        await page.goto('http://localhost:5050', { waitUntil: 'networkidle0' });
        
        // Wait for page to load
        await page.waitForSelector('#new-game-btn', { timeout: 5000 });
        console.log('✓ Page loaded successfully');
        
        // Test 1: Start a new game
        console.log('Testing new game...');
        await page.click('#new-game-btn');
        await page.waitForTimeout(500);
        
        // Check if board is rendered
        const boardCells = await page.$$eval('.tile', tiles => tiles.length);
        console.log(`✓ Board rendered with ${boardCells} tiles`);
        
        // Test 2: Check if OCaml backend is available
        const aiPlayBtn = await page.$('#ai-play-btn');
        const isDisabled = await page.evaluate(btn => btn.disabled, aiPlayBtn);
        
        if (!isDisabled) {
            console.log('✓ AI Play button is enabled (OCaml backend available)');
            
            // Test 3: Let AI make a few moves
            console.log('Testing AI moves...');
            await page.click('#ai-play-btn');
            
            // Wait for AI to make moves and check inference time
            let inferenceTimeFound = false;
            for (let i = 0; i < 10; i++) {
                await page.waitForTimeout(200);
                
                // Check if inference time is displayed
                const inferenceTime = await page.$eval('#inference-time', el => el.textContent);
                if (inferenceTime !== '-') {
                    inferenceTimeFound = true;
                    console.log(`✓ AI move ${i+1}: Inference time = ${inferenceTime} ms`);
                }
                
                // Check score
                const score = await page.$eval('#game-score', el => el.textContent);
                const maxTile = await page.$eval('#game-max-tile', el => el.textContent);
                console.log(`  Score: ${score}, Max tile: ${maxTile}`);
                
                // Check if game is over
                const gameOver = await page.evaluate(() => {
                    return window.isGameOver || false;
                });
                
                if (gameOver) {
                    console.log('✓ Game ended naturally');
                    break;
                }
            }
            
            if (inferenceTimeFound) {
                console.log('✓ OCaml inference times displayed correctly');
            } else {
                console.log('✗ No inference times found - OCaml backend may not be working');
            }
            
            // Stop AI
            await page.click('#ai-stop-btn');
            await page.waitForTimeout(200);
            
        } else {
            console.log('✗ AI Play button is disabled - OCaml backend not available');
        }
        
        // Test 4: Switch to Train tab
        console.log('Testing Train tab...');
        await page.evaluate(() => {
            const trainTab = Array.from(document.querySelectorAll('.tab-link')).find(el => el.textContent === 'Train');
            if (trainTab) trainTab.click();
        });
        await page.waitForTimeout(500);
        
        // Check if training controls are visible
        const startBtn = await page.$('#start-btn');
        if (startBtn) {
            console.log('✓ Train tab loaded successfully');
        }
        
        console.log('\n✅ All tests passed!');
        
    } catch (error) {
        console.error('❌ Test failed:', error.message);
        
        // Take screenshot on error
        await page.screenshot({ path: 'test_error.png' });
        console.log('Screenshot saved as test_error.png');
        
    } finally {
        await browser.close();
    }
}

// Check if required packages are installed
try {
    require.resolve('puppeteer');
} catch(e) {
    console.error('Puppeteer not installed. Run: npm install puppeteer');
    process.exit(1);
}

// Run the test
testWebInterface().catch(console.error);